api/agent/drivers/docker/docker.go

```diff
@@ -368,35 +367,22 @@ func dockerMsg(derr *docker.Error) string {
 // The docker driver will attempt to cast the task to a Auther. If that succeeds, private image support is available. See the Auther interface for how to implement this.
 func (drv *DockerDriver) run(ctx context.Context, container string, task drivers.ContainerTask) (drivers.WaitResult, error) {
 
-	attachSuccess := make(chan struct{})
 	mwOut, mwErr := task.Logger()
-
-	waiter, err := drv.docker.AttachToContainerNonBlocking(ctx, docker.AttachToContainerOptions{
-		Success:      attachSuccess,
+	waiter, err := drv.docker.AttachToContainer(ctx, docker.AttachToContainerOptions{
```

Comment:
this doesn't respect the context afaict https://github.com/fsouza/go-dockerclient/blob/master/container.go#L1441 -- hence the NB attach. for metrics we can probably get them in the waiter, but the waiter dance is critical until this is fixed upstream. have attempted to fix this upstream before https://github.com/moby/moby/pull/36544 (and move to that docker client) but alas, nothing landed. this is a bad bug we need to fix asap, since if the docker client doesn't timeout properly we can leak goroutines and containers... sorry, should have caught this earlier.

Reply:
@skinowski ^ urgent

Reply:
ah, i see we just renamed the thing to try to push it down but it's the same thing. i think still kinda misses the idea of the wrapper around the client (of which i am ashamed in general, but don't see any better options), which was to keep the same behaviors of the underlying fsouza api but add retries and timeouts to the docker calls since docker api calls like to fail frequently... anyway, no bug if underneath uses NB, but this still smells since it's a mismatch